cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,417 @@
+
+	if clientset == nil {
+		klog.Warning("no clientset available, resource manager not created")
+		return nil
+	}
+
+	rm := ResourceManager{
+		nodeName:  os.Getenv("NODE_NAME"),
+		clientset: clientset,
+	}
+
+	klog.Info("GPU device plugin resource manager enabled")
+
+	return &rm
+}
+
+// ReallocateWithFractionalResources runs the fractional resource logic.
+// This intentionally only logs errors and returns with the default device allocation, in case any errors are hit. This is to avoid
+// clusters filling up with unexpected admission errors.
+func (rm *ResourceManager) ReallocateWithFractionalResources(request *pluginapi.AllocateRequest, allocateResponse *pluginapi.AllocateResponse) {
+	var err error
+	if !isInputOk(request, allocateResponse) {
+		// it is better to leave allocated gpu devices as is and return
+		return
+	}
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if errors.Is(err, errRetry) {
+		klog.Warning("retrying POD resolving after sleeping")
+		time.Sleep(retryTimeout)
+		podCandidate, err = rm.findAllocationPodCandidate()
+	}
+
+	if err != nil {
+		if !errors.Is(err, errZeroPending) {
+			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:" +
+				err.Error())
+		}
+		// it is better to leave allocated gpu devices as is and return
+		return
+	}
+
+	pod := podCandidate.pod
+	cards := containerCards(pod, podCandidate.allocatedContainers)
+	logPodInfo(pod, cards)
+
+	cresp := allocateResponse.ContainerResponses[0]
+	// select gpus for the allocation. the cards may differ from kubelet given deviceIDs!
+	for _, cardName := range cards {
+		rm.addDevicesAndMounts(cresp, cardName)
+	}
+}
+
+func isInputOk(rqt *pluginapi.AllocateRequest, allocateResponse *pluginapi.AllocateResponse) bool {
+	// so far kubelet calls allocate for each container separately. If that changes, we need to refine our logic.
+	if len(allocateResponse.ContainerResponses) != 1 {
+		klog.Warning("multi-container allocation request")
+		return false
+	}
+
+	for _, crqt := range rqt.ContainerRequests {
+		for _, id := range crqt.DevicesIDs {
+			if id == monitorID {
+				return false // intentionally not printing anything
+			}
+		}
+	}
+
+	return true
+}
+
+// findAllocationPodCandidate tries to find the best allocation candidate pod, which must be:
+//  -pending for this node
+//  -using GPU resources in its spec
+//  -is found via grpc service with unallocated GPU devices
+// returns:
+//  -the candidate pod struct pointer and no error, or
+//  -errRetry if unsuccessful, but there is perhaps hope of trying again with better luck
+//  -errZeroPending if no pending pods exist anymore (which is fine)
+//  -any grpc communication errors
+func (rm *ResourceManager) findAllocationPodCandidate() (*podCandidate, error) {
+	// get map of pending pods for this node
+	pendingPods, err := rm.getNodePendingGPUPods()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting pending pods failed")
```

Comment:
```suggestion
		return nil, err
```
Usually we wrap errors only if they lack stack trace.